File: PythonCodingTest/백준/단지번호붙이기.py
import sys
sys.setrecursionlimit(10 ** 6)

n= int(input())
arr= []
dx=[0,0,-1,1]
dy=[-1,1,0,0]
dangi= []
size=1

for i in range(n):
  arr.append(list(map(int,input())))
  
def dfs(x,y):
  global size;
  for k in range(4):
    nx= x+dx[k]
    ny= y+dy[k]
    if(nx<0 or ny<0 or nx>=n or ny>= n): continue
    if(arr[nx][ny]==1):
      size+=1
      arr[nx][ny]=2 #들어간 집 체크
      dfs(nx,ny)

for i in range(n):
  for j in range(n):
    if(arr[i][j]==1):
      arr[i][j]=2 #방문한 것으로 체크
      dfs(i,j)
      dangi.append(size) #단지 내 집 수 배열에 추가
      size=1 #사이즈 초기화
      
dangi.sort()
print(len(dangi))
for k in dangi:
  print(k)

File: 백준/그래프/단지번호붙이기.js
const fs = require('fs').readFileSync(
  process.platform === 'linux' ? '/dev/stdin' : __dirname + '/input.txt'
);
const input = fs.toString().trim().split('\n');
const count = Number(input[0]);
const map = input.slice(1).map((el) => el.split('').map(Number));

const dx = [0, 0, 1, -1];
const dy = [1, -1, 0, 0];
let dangi = [];
let cnt = 1;
//dfs 함수만들기
function dfs(x, y) {
  for (let i = 0; i < 4; i++) {
    const nx = x + dx[i];
    const ny = y + dy[i];

    if (nx < 0 || nx >= count || ny < 0 || ny >= count) {
      continue;
    }
    if (map[nx][ny] == 1) {
      map[nx][ny] = 2;
      cnt++;
      dfs(nx, ny);
    }
  }
}
for (let i = 0; i < count; i++) {
  for (let j = 0; j < count; j++) {
    if (map[i][j] == 1) {
      map[i][j] = 2;
      dfs(i, j);
      dangi.push(cnt);
      cnt = 1;
    }
  }
}
console.log(dangi.length);
dangi.sort((a, b) => a - b);
dangi.forEach((n) => console.log(n));

//output: 총 단지 수, 각 단지별 세대 수. =>DFS
